### DISCUSSION TITLE
createSchemaCustomization hangs forever

### DISCUSSION THREAD

**Discussion by abemedia** [gatsbyjs/gatsby#36133] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I recently migrated my `gatsby-node` file to Typescript, and to satisfy the type I had to add the `callback` argument. This caused the `createSchemaCustomization` function to hang forever.

Before my file looked as follows and worked fine:

```js
const content = require('./internal/content')
const downloads = require('./internal/downloads')

exports.createSchemaCustomization = props => {
  content.createSchemaCustomization(props)
  downloads.createSchemaCustomization(props)
}
```

Then, as the imported functions were given the type `GatsbyNode['createSchemaCustomization']`, I needed to pass `callback` to them, so the file became the following and the build hung forever:

```ts
import type { GatsbyNode } from 'gatsby'
import * as content from './internal/content'
import * as downloads from './internal/downloads'

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = (
  args, 
  options, 
  callback
) => {
  content.createSchemaCustomization?.(args, options, callback)
  downloads.createSchemaCustomization?.(args, options, callback)
}
```

I managed to fix it by changing the function to use the spread operator:

```ts
export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = (...args) => {
  content.createSchemaCustomization?.(...args)
  downloads.createSchemaCustomization?.(...args)
}
```

I think the `callback` argument should be made optional in the type to avoid this.

### Reproduction Link

n/a

### Steps to Reproduce

1. Add a `createSchemaCustomization` function to `gatsby-node` which has an (unused) callback parameter.
2. Run `gatsby develop` or `gatsby build`

### Expected Result

Gatsby builds as usual.

### Actual Result

Build hung at `createSchemaCustomization` forever.

### Environment

```shell
System:
    OS: Linux 5.4 Ubuntu 18.04.6 LTS (Bionic Beaver)
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 16.13.2 - /tmp/yarn--1657829502293-0.3151360212459857/node
    Yarn: 1.22.19 - /tmp/yarn--1657829502293-0.3151360212459857/yarn
    npm: 8.1.2 - ~/n/bin/npm
  Languages:
    Python: 2.7.17 - /usr/bin/python
  Browsers:
    Chrome: 103.0.5060.114
    Firefox: 102.0
  npmPackages:
    gatsby: ^4.18.0 => 4.18.0 
    gatsby-link: ^4.16.0 => 4.18.0 
    gatsby-plugin-babel-plugin-typescript-to-proptypes: ^1.0.5 => 1.0.5 
    gatsby-plugin-catch-links: ^4.17.0 => 4.17.0 
    gatsby-plugin-google-analytics: ^4.17.0 => 4.17.0 
    gatsby-plugin-image: ^2.18.0 => 2.18.0 
    gatsby-plugin-manifest: ^4.18.0 => 4.18.0 
    gatsby-plugin-meta-redirect: ^1.1.1 => 1.1.1 
    gatsby-plugin-offline: ^5.17.0 => 5.17.0 
    gatsby-plugin-react-helmet: ^5.17.0 => 5.17.0 
    gatsby-plugin-robots-txt: ^1.7.1 => 1.7.1 
    gatsby-plugin-sass: ^5.17.0 => 5.17.0 
    gatsby-plugin-sharp: ^4.17.0 => 4.17.0 
    gatsby-plugin-sitemap: ^5.17.0 => 5.17.0 
    gatsby-plugin-smoothscroll: ^1.1.0 => 1.2.0 
    gatsby-plugin-typescript-checker: ^1.1.1 => 1.1.1 
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.27 => 1.1.27 
    gatsby-redirect-from: ^0.5.0 => 0.5.0 
    gatsby-remark-autolink-headers: ^5.17.0 => 5.17.0 
    gatsby-remark-external-links: ^0.0.4 => 0.0.4 
    gatsby-remark-images: ^6.18.0 => 6.18.0 
    gatsby-remark-smartypants: ^5.17.0 => 5.17.0 
    gatsby-remark-vscode: ^3.3.1 => 3.3.1 
    gatsby-source-filesystem: ^4.17.0 => 4.17.0 
    gatsby-transformer-inline-svg: ^1.2.0 => 1.2.0 
    gatsby-transformer-remark: ^5.18.0 => 5.18.0 
    gatsby-transformer-sharp: ^4.18.0 => 4.18.0 
    gatsby-transformer-yaml: ^4.18.0 => 4.18.0
```

**Comment by LekoArts** [comment, marked as answer]
All three parameters are optional and you don't have to define a callback. This here works:

```ts
import { GatsbyNode, CreateSchemaCustomizationArgs } from "gatsby"

const doCustomStuff = (args: CreateSchemaCustomizationArgs) => {
  args.actions.createTypes(``)
}

export const createSchemaCustomization: GatsbyNode["createSchemaCustomization"] = (args) => {
  doCustomStuff(args)
}
```

What these parameters do is explained here: https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#async-vs-sync-work

The last relevant change to the types was made here: https://github.com/gatsbyjs/gatsby/pull/30819

> Each of the GatsbyNode apis can be characterized with all three parameters: args, options, and the callback. Whoever is writing the endpoint can choose which parameters they wish to consume.

**Reply by abemedia** [reply]
Thanks for your response. They aren't optional in `GatsbyNode["createSchemaCustomization"]` though.

Your solution is definitely cleaner looking than the spread operator like I used, however since the parameters are in fact optional the type should reflect this.

I'm happy to raise a PR if this is acceptable.

**Reply by LekoArts** [reply]
The type is for the main export, for any helper functions you create you'll need to use the proper types like `CreateSchemaCustomizationArgs`. So the types don't need fixing
